operator/builtin/input/aws/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,376 @@
+// and returns completed input.
+func (c *CloudwatchInput) filterLogEventsInputBuilder(nextToken string) cloudwatchlogs.FilterLogEventsInput {
+	logGroupNamePtr := aws.String(c.logGroupName)
+	limit := aws.Int64(c.eventLimit)
+	startTime := aws.Int64(c.startTime)
+
+	if c.logStreamNamePrefix != "" && nextToken != "" {
+		tmp := timeLayoutParser(c.logStreamNamePrefix, time.Now())
+		logStreamNamePrefixPtr := aws.String(tmp)
+		nextTokenPtr := aws.String(nextToken)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:               limit,
+			LogGroupName:        logGroupNamePtr,
+			LogStreamNamePrefix: logStreamNamePrefixPtr,
+			StartTime:           startTime,
+			NextToken:           nextTokenPtr,
+		}
+	}
+
+	if c.logStreamNamePrefix != "" {
+		tmp := timeLayoutParser(c.logStreamNamePrefix, time.Now())
+		logStreamNamePrefixPtr := aws.String(tmp)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:               limit,
+			LogGroupName:        logGroupNamePtr,
+			LogStreamNamePrefix: logStreamNamePrefixPtr,
+			StartTime:           startTime,
+		}
+	}
+
+	if len(c.logStreamNames) > 0 && nextToken != "" {
+		nextTokenPtr := aws.String(nextToken)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:          limit,
+			LogGroupName:   logGroupNamePtr,
+			LogStreamNames: c.logStreamNames,
+			StartTime:      startTime,
+			NextToken:      nextTokenPtr,
+		}
+	}
+
+	if len(c.logStreamNames) > 0 {
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:          limit,
+			LogGroupName:   logGroupNamePtr,
+			LogStreamNames: c.logStreamNames,
+			StartTime:      startTime,
+		}
+	}
+
+	if nextToken != "" {
+		nextTokenPtr := aws.String(nextToken)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:        limit,
+			LogGroupName: logGroupNamePtr,
+			StartTime:    startTime,
+			NextToken:    nextTokenPtr,
+		}
+	}
+
+	return cloudwatchlogs.FilterLogEventsInput{
+		Limit:        limit,
+		LogGroupName: logGroupNamePtr,
+		StartTime:    startTime,
+	}
+}
+
+// handleEvent is the handler for a AWS Cloudwatch Logs Filtered Event.
+func (c *CloudwatchInput) handleEvent(ctx context.Context, event *cloudwatchlogs.FilteredLogEvent) error {
+	e := make(map[string]interface{})
+	e["message"] = event.Message
+	e["event_id"] = event.EventId
+	e["ingestion_time"] = event.IngestionTime
+
+	entry, err := c.NewEntry(nil)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create new entry from record")
+	}
+
+	entry.AddResourceKey("log_group_name", c.logGroupName)
+	entry.AddResourceKey("region", c.region)
+	entry.AddResourceKey("log_stream_name", *event.LogStreamName)
+	entry.Timestamp = fromUnixMilli(*event.Timestamp)
+	entry.Record = e
+
+	// Persist
+	if *event.IngestionTime > c.startTime {
+		c.startTime = *event.IngestionTime
+		c.Debugf("Writing start time %d to database", *event.IngestionTime)
+		c.persist.Write(c.logGroupName, *event.IngestionTime)
+	}
```

Comment:
This logic warrants a comment. What's the idea behind this?

Reply:
I will add a comment. This logic is intended to reduce the number of writes. Often in testing I found that batches of events would have the same ingestion time. So we were making unnecessary writes when the startTime was the same as the ingestion time.

Reply:
Typically, please allow the commenter to resolve conversations.

I still found this confusing, but I think I've got it now. 

I don't think the comment really explains the _idea behind_ the logic. Here's what I suggest:

```go
// Keep track of which events have been consumed, in case of restart
if *event.IngestionTime > c.startTime {
	c.startTime = *event.IngestionTime
	c.Debugf("Writing start time %d to database", *event.IngestionTime)
	c.persist.Write(c.logGroupName, c.startTime)
}
```

Note that even though the values are the same, it is more readable to write `c.startTime`, since that's the field that keeps track of where we need to start.

Reply:
Made change.